### PR TITLE
light client upload file support -V3 gateway

### PR DESCRIPTION
**Pull request by aligeti**
Integrated eestream's serve-piece functionality

### CODE REVIEW

File: pkg/miniogw/gateway-storj.go [hunk 1]
```diff
@@ -5,27 +5,196 @@ package storj
...
+}
+
+//S3FileList structure
+type S3FileList struct {
+	totalFiles int
+	file       minio.ListObjectsInfo
+}
+
+// gS3Gateway global S3 interface structure
+var gS3CliAPI S3CliAPI
```
Comment (jtolds) [inline pkg/miniogw/gateway-storj.go]: i'd really love to avoid checking in any global variables here. i know we talked about this before, think that could be within scope for this pull request?

File: pkg/miniogw/gateway-storj.go [hunk 2]
```diff
@@ -5,27 +5,196 @@ package storj
...
+
+//createDummyBucketList function initializes sample buckets and files in each bucket
+func createDummyBucketList() {
+	gS3CliAPI.bucketlist = gS3CliAPI.bucketlist[:0]
```
[the remark this answers is not in the document]
Comment (jtolds) [reply pkg/miniogw/gateway-storj.go]: it's resetting the global state. i think we should avoid global state and then we don't need this

File: pkg/miniogw/gateway-storj.go [hunk 3]
```diff
@@ -5,27 +5,196 @@ package storj
...
-	cmd.RegisterGatewayCommand(cli.Command{
+	minio.RegisterGatewayCommand(cli.Command{
 		Name:            "storj",
 		Usage:           "Storj",
 		Action:          storjGatewayMain,
 		HideHelpCommand: true,
 	})
+
+	// create dummy bucket list
+	createDummyBucketList()
```
Comment (jtolds) [inline pkg/miniogw/gateway-storj.go]: instead of having this function mutate global state, you can have the Action on line 94 have a "closure" which is a function that is lexically bound to some specific state.

for instance, this could be:

```
func init() {
   s := &Storj{}
   s.createDummyBucketList()
   minio.RegisterGatewayCommand(cli.Command{
     Name: "storj",
     Usage: "Storj",
     Action: func(ctx *cli.Context) {
        minio.StartGateway(ctx, s)
     },
     HideHelpCommand: true,
   })
}
```

notice how `createDummyBucketList` is defined on a specific struct, thus not needing a global.

File: pkg/miniogw/gateway-storj.go [hunk 4]
```diff
@@ -5,27 +5,196 @@ package storj
...
+}
+
+// getbucketscallback
+func getbucketscallback() (buckets []minio.BucketInfo, err error) {
```
Comment (jtolds) [inline pkg/miniogw/gateway-storj.go]: if we're going to eliminate global state this needs to be defined on a struct, like

```
func (s *Storj) getbucketscallback() (buckets []minio.BucketInfo, err error) {
```

then you can use the state from `s` instead of a global

File: pkg/miniogw/gateway-storj.go [hunk 5]
```diff
@@ -5,27 +5,196 @@ package storj
...
+			Created: bi.bucket.Created,
+		}
+	}
+	fmt.Println("buckets list =", buckets)
+	return buckets, nil
+}
+
+// putobjectcallback
+func putobjectcallback(bucket, object string, filesize int64, metadata map[string]string) (result minio.ListObjectsInfo, err error) {
+	fmt.Printf("Go.putobjectcallback(): called \n")
```
Comment (jtolds) [inline pkg/miniogw/gateway-storj.go]: same as kaloyan's comment about fmt.Print*. better to use a logger

File: pkg/miniogw/gateway-storj.go [hunk 6]
```diff
@@ -72,48 +242,106 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
 }
 
 func (s *storjObjects) ListBuckets(ctx context.Context) (
-	buckets []cmd.BucketInfo, err error) {
-	return []cmd.BucketInfo{{
-		Name:    "test-bucket",
-		Created: time.Now(),
-	}}, nil
+	buckets []minio.BucketInfo, err error) {
+	buckets, err = getbucketscallback()
```
Comment (jtolds) [inline pkg/miniogw/gateway-storj.go]: so here, you should be able to call s.s.getbucketscallback() instead, or just put all that logic in here. that way it's defined on a struct and you don't need global memory

File: pkg/miniogw/gateway-storj.go [hunk 7]
```diff
@@ -48,7 +217,8 @@ func (s *Storj) Production() bool {
 }
 
 type storjObjects struct {
-	cmd.GatewayUnsupported
+	minio.GatewayUnsupported
+	TempDir string // Temporary storage location for file transfers.
```
Comment (jtolds) [inline pkg/miniogw/gateway-storj.go]: also maybe have a reference to `*Storj` here, like

```
type storjObjects struct {
   minio.GatewayUnsupported
   TempDir string
   s *Storj
}
```

File: pkg/miniogw/gateway-storj.go [hunk 8]
```diff
@@ -38,7 +207,7 @@ func (s *Storj) Name() string {
 
 // NewGatewayLayer implements cmd.Gateway
 func (s *Storj) NewGatewayLayer(creds auth.Credentials) (
-	cmd.ObjectLayer, error) {
+	minio.ObjectLayer, error) {
```
Comment (jtolds) [inline pkg/miniogw/gateway-storj.go]: see the comment below, but we can do `return &storjObjects{s:s}` to pass in the non-global state

File: pkg/miniogw/gateway-storj.go [hunk 9]
```diff
@@ -72,48 +242,106 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
-			Bucket:      "test-bucket",
-			Name:        "test-file",
-			ModTime:     time.Now(),
-			Size:        0,
-			IsDir:       false,
-			ContentType: "application/octet-stream",
-		}},
-	}, nil
+	delimiter string, maxKeys int) (result minio.ListObjectsInfo, err error) {
+	result, err = listfilescallback(bucket)
```
Comment (jtolds) [inline pkg/miniogw/gateway-storj.go]: same. why not have all of the logic in here instead of inside another function? also, that way you have access to `s.s`

File: pkg/miniogw/gateway-storj.go [hunk 10]
```diff
@@ -5,27 +5,196 @@ package storj
...
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
 	"time"
 
 	"github.com/minio/cli"
+	"github.com/vivint/infectious"
+	"storj.io/storj/pkg/eestream"
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Just curious, I know of reading interfaces to be named after the methods like reader because of read method. But haven't come across to leave an empty line from the GitHub.com group imports.... very interesting... will make the change.

File: pkg/miniogw/gateway-storj.go [hunk 11]
```diff
@@ -5,27 +5,196 @@ package storj
...
+	file       minio.ListObjectsInfo
+}
+
+// gS3Gateway global S3 interface structure
+var gS3CliAPI S3CliAPI
+
+//createDummyBucketList function initializes sample buckets and files in each bucket
+func createDummyBucketList() {
+	gS3CliAPI.bucketlist = gS3CliAPI.bucketlist[:0]
+	for i := 0x00; i < 0x0A; i++ {
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: I didn't even realized that...:-) old habits die hard....from system programming for decades... will do the mod

File: pkg/miniogw/gateway-storj.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: It is just init code. This function creates a dummy bucket and files list in each bucket. This whole function will be removed once we start integrating with high clients. If I understand it correctly, once we have high clients up and running, the storj gateway will be receiving the info from high clients which would be based on the specific requests coming from S3. If you guys really insist, I will make the change but if you look at the big picture, it is just a init function, which fill in data, with a variable scoped only to the gateway-storj file. It cannot be accessed outside this file.

File: pkg/miniogw/gateway-storj.go [hunk 12]
```diff
@@ -5,27 +5,196 @@ package storj
...
+			S3BucketList{
+				bucket: minio.BucketInfo{
+					Name:    "TestBucket" + strconv.Itoa(i+1),
+					Created: time.Now(),
+				},
+			},
+		)
+	} /* end of for loop */
+
+	for i := 0x00; i < 0x0A; i++ {
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Absolutely... will do

File: pkg/miniogw/gateway-storj.go [hunk 13]
```diff
@@ -5,27 +5,196 @@ package storj
...
 		Usage:           "Storj",
 		Action:          storjGatewayMain,
 		HideHelpCommand: true,
 	})
+
+	// create dummy bucket list
+	createDummyBucketList()
+}
+
+// getbucketscallback
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: That was there to clear the warnings. will update the comment

File: pkg/miniogw/gateway-storj.go [hunk 14]
```diff
@@ -5,27 +5,196 @@ package storj
...
+					Name:        gS3CliAPI.bucketlist[i].bucket.Name + "file" + strconv.Itoa(j+1),
+					ModTime:     time.Now(),
+					Size:        int64(100 + i),
+					ContentType: "application/octet-stream",
+				},
+			)
+		}
+	}
+	gS3CliAPI.totalBuckets = len(gS3CliAPI.bucketlist)
+	fmt.Println("bucket name = ", gS3CliAPI.bucketlist)
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: used it for debugging the code. Will clean it up.

**Comment by kaloyan-raev** [comment]
After some debugging together with Kishore we found that minio cmd requires that MINIO_ACCESS_KEY and MINIO_SECRET_KEY env vars are set. I can now test the storj gateway.

File: pkg/miniogw/gateway-storj.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Removed the global references and implemented as suggested

File: pkg/miniogw/gateway-storj.go [hunk 15]
```diff
@@ -5,27 +5,196 @@ package storj
...
+		}
+	}
+	fmt.Println("buckets list =", buckets)
+	return buckets, nil
+}
+
+// putobjectcallback
+func putobjectcallback(bucket, object string, filesize int64, metadata map[string]string) (result minio.ListObjectsInfo, err error) {
+	fmt.Printf("Go.putobjectcallback(): called \n")
+	var bucketName string
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: done

File: pkg/miniogw/gateway-storj.go [hunk 16]
```diff
@@ -5,27 +5,196 @@ package storj
...
+	var fl []minio.ObjectInfo
+	for i, v := range gS3CliAPI.bucketlist {
+		bucketName = v.bucket.Name
+		ret := strings.Compare(bucketName, bucket)
+		if ret == 0x00 {
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: done

File: pkg/miniogw/gateway-storj.go (hunk 1 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Modified as suggested to eliminate global references

File: pkg/miniogw/gateway-storj.go (hunk 3 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Modified as suggested to eliminate global references

File: pkg/miniogw/gateway-storj.go (hunk 4 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Modified as suggested to eliminate global references

File: pkg/miniogw/gateway-storj.go (hunk 5 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Cleaned out the debug information

File: pkg/miniogw/gateway-storj.go (hunk 6 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Modified as suggested to eliminate global references

File: pkg/miniogw/gateway-storj.go (hunk 7 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Modified as suggested to eliminate global references

File: pkg/miniogw/gateway-storj.go (hunk 8 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Modified as suggested to eliminate global references

File: pkg/miniogw/gateway-storj.go (hunk 9 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: Modified as suggested to eliminate global references

File: pkg/miniogw/gateway-storj.go [hunk 17]
```diff
@@ -38,17 +152,49 @@ func (s *Storj) Name() string {
...
 }
 
 // Production implements cmd.Gateway
 func (s *Storj) Production() bool {
 	return false
 }
 
+//createSampleBucketList function initializes sample buckets and files in each bucket
+func (s *Storj) createSampleBucketList() {
+	s.bucketlist = s.bucketlist[:0]
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: What does this line do?

File: pkg/miniogw/gateway-storj.go [hunk 18]
```diff
@@ -72,48 +218,102 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
 }
 
 func (s *storjObjects) ListBuckets(ctx context.Context) (
-	buckets []cmd.BucketInfo, err error) {
-	return []cmd.BucketInfo{{
-		Name:    "test-bucket",
-		Created: time.Now(),
-	}}, nil
+	buckets []minio.BucketInfo, err error) {
+	buckets, err = s.getbucketscallback()
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: I don't see any benefit of extracting the logic in a private method. Perhaps, just move it back here inline.

File: pkg/miniogw/gateway-storj.go [hunk 19]
```diff
@@ -5,31 +5,151 @@ package storj
...
+	"github.com/vivint/infectious"
+
+	"storj.io/storj/pkg/eestream"
+)
+
+var (
+	pieceBlockSize = flag.Int("piece_block_size", 4*1024, "block size of pieces")
+	key            = flag.String("key", "a key", "the secret key")
+	rsk            = flag.Int("required", 20, "rs required")
+	rsn            = flag.Int("total", 40, "rs total")
```
[the remark this answers is not in the document]
Comment (jtolds) [reply pkg/miniogw/gateway-storj.go]: seems fine to me

File: pkg/miniogw/gateway-storj.go (hunk 17 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: it initializes the slice.  Is that a comment or a clarification ?

File: pkg/miniogw/gateway-storj.go [hunk 20]
```diff
@@ -72,48 +218,102 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
+
+	wsize, err := io.CopyN(writer, data, data.Size())
+	if err != nil {
+		os.Remove(srcFile)
+		return objInfo, err
+	}
+
+	uploadcallback(writer, uint(wsize), bucket, object)
+
+	_, _ = s.putobjectcallback(bucket, object, wsize, metadata)
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: s.putobjectcallback this function just appends the uploaded file to the list. I added the return values for debugging purposes.

File: pkg/miniogw/gateway-storj.go [hunk 21]
```diff
@@ -5,31 +5,145 @@ package storj
...
-	cmd.RegisterGatewayCommand(cli.Command{
+	minio.RegisterGatewayCommand(cli.Command{
 		Name:            "storj",
 		Usage:           "Storj",
 		Action:          storjGatewayMain,
 		HideHelpCommand: true,
 	})
 }
 
+// getbucketscallback returns the buckets list
```
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: you right...will settle with getFiles, uploadFile, getBuckets, ...makes more sense

File: pkg/miniogw/gateway-storj.go (hunk 17 above)
Comment (kaloyan-raev) [reply pkg/miniogw/gateway-storj.go]: IMHO, this is a pretty confusing way for initializing an empty slice. Since you know the expected size of the slice (10), it would be much better to initialize it like `s.bucketlist = make([]S3BucketList, 10)`. This will avoid resizing the internal slice buffer when adding items. Even better, later in the for-loop you can get rid of the `append` and set the items directly at the designated index, i.e. `s.bucketlist[i] = S3BucketList{...}`.

File: pkg/miniogw/gateway-storj.go (hunk 18 above)
Comment (kaloyan-raev) [reply pkg/miniogw/gateway-storj.go]: Then you can do it just like `return s.getbucketscallback()`.

File: pkg/miniogw/gateway-storj.go [hunk 22]
```diff
@@ -72,48 +218,102 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
-			Name:        "test-file",
-			ModTime:     time.Now(),
-			Size:        0,
-			IsDir:       false,
-			ContentType: "application/octet-stream",
-		}},
-	}, nil
+	delimiter string, maxKeys int) (result minio.ListObjectsInfo, err error) {
+	result, err = s.listfilescallback(bucket)
+	return result, nil
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: Same here: `return s.listfilescallback(bucket)`, which will also fix the bug of losing the error :)

File: pkg/miniogw/gateway-storj.go [hunk 23]
```diff
@@ -72,48 +218,102 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
 }
 
 func (s *storjObjects) MakeBucketWithLocation(ctx context.Context,
 	bucket string, location string) error {
 	panic("TODO")
 }
 
+//uploadcallback uploads the files
+func uploadcallback(data io.ReadCloser, blockSize uint, bucket, object string) error {
+	dir := "/tmp/gateway/" + bucket + "/" + object
```
[the remark this answers is not in the document]
Comment (kaloyan-raev) [reply pkg/miniogw/gateway-storj.go]: I didn't get that...
Also note that `/tmp/gateway` won't work on Windows.

File: pkg/miniogw/gateway-storj.go (hunk 18 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: :-) I see... well another Misra coding style from the embedded system programming world....I see your point.. wilco ( will comply)

File: pkg/miniogw/gateway-storj.go (hunk 22 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: wilco

File: pkg/miniogw/gateway-storj.go (hunk 17 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: wilco

Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: simplified the function (rewritten)

File: pkg/miniogw/gateway-storj.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: used os.TempDir() to get the temp directory

File: pkg/miniogw/gateway-storj.go [hunk 24]
```diff
@@ -72,48 +204,104 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
 
 func (s *storjObjects) MakeBucketWithLocation(ctx context.Context,
 	bucket string, location string) error {
 	panic("TODO")
 }
 
+//encryptFile encrypts the uploaded files
+func encryptFile(data io.ReadCloser, blockSize uint, bucket, object string) error {
+	dir := os.TempDir()
+	dir = dir + "gateway/" + bucket + "/" + object
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: `filepath.Join()` should be a more reliable way of creating file paths: https://golang.org/pkg/path/filepath/#Join

File: pkg/miniogw/gateway-storj.go [hunk 25]
```diff
@@ -72,48 +204,104 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
+	fmt.Println("temp dir = ", dir)
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: Don't forget to remove this before committing, or use the log package if that log message is important in future.

File: pkg/miniogw/gateway-storj.go [hunk 26]
```diff
@@ -72,48 +204,104 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
+	wsize, err := io.CopyN(writer, data, data.Size())
+	if err != nil {
+		os.Remove(srcFile)
+		return objInfo, err
+	}
+
+	err = encryptFile(writer, uint(wsize), bucket, object)
+	if err == nil {
+		// return values are for debugging purposes
+		_, _ = s.uploadFile(bucket, object, wsize, metadata)
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: I don't see these return values used for anything. Perhaps, remove them together with the comment.

File: pkg/miniogw/gateway-storj.go [hunk 27]
```diff
@@ -5,31 +5,143 @@ package storj
...
+	minio.RegisterGatewayCommand(cli.Command{
 		Name:            "storj",
 		Usage:           "Storj",
 		Action:          storjGatewayMain,
 		HideHelpCommand: true,
 	})
 }
 
+// getbucketscallback returns the buckets list
+func (s *storjObjects) getBuckets() (buckets []minio.BucketInfo, err error) {
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: This method does not seems to have a case where an error is returned. Thus, why having and error as a return value?

Comment (kaloyan-raev) [reply pkg/miniogw/gateway-storj.go]: Same for uploadFile() and getFiles()

File: pkg/miniogw/gateway-storj.go [hunk 28]
```diff
@@ -5,31 +5,143 @@ package storj
...
+	s := &Storj{}
+	s.createSampleBucketList()
+	minio.StartGateway(ctx, s)
+}
+
+//S3BucketList structure
+type S3BucketList struct {
+	bucket   minio.BucketInfo
+	filelist S3FileList
+}
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: As far as I understand an instance of this type represents a single bucket, right? So, why is the "List" suffix in the name?

File: pkg/miniogw/gateway-storj.go [hunk 29]
```diff
@@ -5,31 +5,143 @@ package storj
...
+
+//S3FileList structure
+type S3FileList struct {
+	file minio.ListObjectsInfo
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: I don't see the benefit of this type. Can't the minio.ListObjectsInfo field be inline the S3BucketList? This will make the code for accessing the objects a little bit shorter.

File: pkg/miniogw/gateway-storj.go [hunk 30]
```diff
@@ -5,31 +5,143 @@ package storj
...
 }
 
 // Storj is the implementation of a minio cmd.Gateway
-type Storj struct{}
+type Storj struct {
+	bucketlist []S3BucketList
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: Why not just `buckets` instead of `bucketlist`? The shorter, the better.

File: pkg/miniogw/gateway-storj.go [hunk 31]
```diff
@@ -5,31 +5,143 @@ package storj
...
+}
+
+// listfilescallback returns the files list for a bucket
+func (s *storjObjects) getFiles(bucket string) (result minio.ListObjectsInfo, err error) {
+	var fl []minio.ObjectInfo
+	for i, v := range s.storj.bucketlist {
+		if v.bucket.Name == bucket {
+			/* populate the filelist */
+			f := make([]minio.ObjectInfo, len(s.storj.bucketlist[i].filelist.file.Objects))
+			for j, fi := range s.storj.bucketlist[i].filelist.file.Objects {
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: I don't understand why you are recreating the ListObjectInfo here. Isn't `s.storj.bucketlist[i].filelist.file.Objects` exactly what has to be returned?

File: pkg/miniogw/gateway-storj.go [hunk 32]
```diff
@@ -5,31 +5,143 @@ package storj
...
+					ModTime:     time.Now(),
+					Size:        filesize,
+					IsDir:       false,
+					ContentType: "application/octet-stream",
+				},
+			)
+			/* populate the filelist */
+			f := make([]minio.ObjectInfo, len(s.storj.bucketlist[i].filelist.file.Objects))
+			for j, fi := range s.storj.bucketlist[i].filelist.file.Objects {
+				f[j] = minio.ObjectInfo{
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: After appending the new file, `s.storj.bucketlist[i].filelist.file.Objects` is what has to be returned, no? Why recreating it?

File: pkg/miniogw/gateway-storj.go (hunk 25 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: That I missed to take it out. I will remove it

File: pkg/miniogw/gateway-storj.go (hunk 28 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: That is because there is a file list object inside S3FileList structure and primarily used for listing files. No change will be made

File: pkg/miniogw/gateway-storj.go (hunk 29 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: I prefer to design the things with future implementations design thoughts in mind. Having said that, I have a strong suspicion that more fields will be added as we integrate with high clients. Also most of the code that has hardcoded/dummy init stuff will be removed. So no change will be made

File: pkg/miniogw/gateway-storj.go (hunk 30 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: it holds the list of buckets... thats how I visioned. No change will be made

File: pkg/miniogw/gateway-storj.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: done

File: pkg/miniogw/gateway-storj.go (hunk 24 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: I will look into it.. no change for now

File: pkg/miniogw/gateway-storj.go (hunk 26 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: done

File: pkg/miniogw/gateway-storj.go (hunk 24 above)
Comment (kaloyan-raev) [reply pkg/miniogw/gateway-storj.go]: I guess you don't see the bug you have in this line. You can easily avoid it by just using filepath.Join(), which by the way you already use at another place in your PR.

File: pkg/miniogw/gateway-storj.go (hunk 27 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: I have it for future development when we start integrating with high client, you will need to have the error conditions reported and handled. No change

File: pkg/miniogw/gateway-storj.go (hunk 32 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: I understand, I kept it for debugging purpose, as I mentioned. I wanted to see if it populates the files list. this is will change once we start integrating with high client. So no change will be made

File: pkg/miniogw/gateway-storj.go (hunk 31 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: you can code this in one line. The implementation I did is to support the future enhancement. Technically I understand all this can be one line and you don't even need this function. I will keep it there for now till, all the features are supported. No change will be made

File: pkg/miniogw/gateway-storj.go (hunk 28 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: S3BucketList hold buckets (list of buckets) and inside each bucket there is a list of files.

File: pkg/miniogw/gateway-storj.go (hunk 24 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: I still not seeing the bug. Interested to know where does it fail and why. I will make the change, but love to learn it. let me know ... I will call you later

Comment (kaloyan-raev) [reply pkg/miniogw/gateway-storj.go]: Check here: https://play.golang.org/p/4uhcPTzPil0

File: pkg/miniogw/gateway-storj.go (hunk 28 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: I now understand.. ..thanks...good point...I am on it

File: pkg/miniogw/gateway-storj.go (hunk 29 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: one reason I kept it there because if you need to store any specific metric pertaining to the list of files particular to a bucket, this is the place holder. bit of future thought based on some experience in the past. For now I would recommend to leave it as is.

File: pkg/miniogw/gateway-storj.go (hunk 24 above)
Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: yep agreed...... done implemented the change

File: pkg/miniogw/gateway-storj.go [hunk 33]
```diff
@@ -72,48 +204,106 @@ func (s *storjObjects) GetObject(ctx context.Context, bucket, object string,
...
+	es := eestream.NewRSScheme(fc, *pieceBlockSize)
+	encKey := sha256.Sum256([]byte(*key))
+	var firstNonce [12]byte
+	encrypter, err := eestream.NewAESGCMEncrypter(
+		&encKey, &firstNonce, es.DecodedBlockSize())
+	if err != nil {
+		return err
+	}
+	readers, err := eestream.EncodeReader(context.Background(), eestream.TransformReader(
+		eestream.PadReader(os.Stdin, encrypter.InBlockSize()), encrypter, 0),
```
Comment (kaloyan-raev) [inline pkg/miniogw/gateway-storj.go]: Why did you change reading from `data` to `os.Stdin`? This broke the upload functionality.

Comment (aligeti) [reply pkg/miniogw/gateway-storj.go]: ooo...man... merge overlook...thnx....it happened because I had to merge the latest changes and the function call's parameters were changed and I had to pull in those changes...well hope this is the last change.

**Comment by kaloyan-raev** [comment]
Approved. The current state is a good first step in implementing file upload in the Minio Gateway.

However, there is still something missing for uploading large files via the AWS CLI. Here is what I get when trying to upload ~200 MB file.
```
aws --endpoint-url http://localhost:9000 s3 cp ~/Videos/ben.mkv s3://TestBucket1/
upload failed: ../../../../../../Videos/ben.mkv to s3://TestBucket1/ben.mkv An error occurred (NotImplemented) when calling the CreateMultipartUpload operation: A header you provided implies functionality that is not implemented
```
But perhaps this can be covered in a new PR.

**Comment by aligeti** [comment]
Thanks for the review...one down...many to GO...
